protocols/imap/src/main/java/org/apache/james/imap/processor/SetAnnotationProcessor.java

```diff
@@ -0,0 +1,84 @@
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.imap.processor;
+
+import java.util.List;
+
+import org.apache.james.imap.api.ImapCommand;
+import org.apache.james.imap.api.ImapConstants;
+import org.apache.james.imap.api.ImapSessionUtils;
+import org.apache.james.imap.api.display.HumanReadableText;
+import org.apache.james.imap.api.message.response.StatusResponse;
+import org.apache.james.imap.api.message.response.StatusResponseFactory;
+import org.apache.james.imap.api.process.ImapProcessor;
+import org.apache.james.imap.api.process.ImapSession;
+import org.apache.james.imap.message.request.SetAnnotationRequest;
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.MailboxManager.MailboxCapabilities;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.slf4j.Logger;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+
+public class SetAnnotationProcessor extends AbstractMailboxProcessor<SetAnnotationRequest> implements CapabilityImplementingProcessor {
+    private final boolean annotationCapabilitySupported;
+
+    public SetAnnotationProcessor(ImapProcessor next, MailboxManager mailboxManager, StatusResponseFactory factory) {
+        super(SetAnnotationRequest.class, next, mailboxManager, factory);
+        annotationCapabilitySupported = mailboxManager.hasCapability(MailboxCapabilities.Annotation);
+    }
+
+    public List<String> getImplementedCapabilities(ImapSession session) {
+        if (annotationCapabilitySupported) {
+            return Lists.newArrayList(ImapConstants.SUPPORTS_ANNOTATION);
+        } else {
+            return ImmutableList.of();
+        }
+    }
+
+    protected void doProcess(SetAnnotationRequest message, ImapSession session, String tag, ImapCommand command,
+            Responder responder) {
+        final MailboxManager mailboxManager = getMailboxManager();
+        final MailboxSession mailboxSession = ImapSessionUtils.getMailboxSession(session);
+        final String mailboxName = message.getMailboxName();
+        try {
+            MailboxPath mailboxPath = buildFullPath(session, mailboxName);
+
+            boolean mailboxExists = mailboxManager.mailboxExists(mailboxPath, mailboxSession);
+
+            if (!mailboxExists) {
+                no(command, tag, responder, HumanReadableText.FAILURE_NO_SUCH_MAILBOX, StatusResponse.ResponseCode.tryCreate());
+            } else {
+                mailboxManager.updateAnnotations(mailboxPath, mailboxSession, Lists.newArrayList(message.getMailboxAnnotation()));
+
+                okComplete(command, tag, responder);
+            }
+        } catch (MailboxException e) {
+            Logger log = session.getLog();
+            if (log.isInfoEnabled()) {
```

Comment:
I think most of the IMAP implementation do this, to ensure to not do useless string concatenation. But this happens only on error handling, I would say it is early optimisation.

I would prefer a session.getLog().info() directly.